File: Windows - console/Lm2.cpp
#include "Lm2.h"
#include "Lm2Command.h"
#include "Variable.h"
#include "LmCommands.h"
#include "InterpreterException.h"
#include "Converter.h"

const int MAX_MEMORY_SIZE = 65535;
const int
	CMD_COMPARE                   = 5,
	CMD_INPUT                     = 6,
	CMD_OUTPUT                    = 7,
	CMD_UNSIGNED_INPUT            = 22,
	CMD_UNSIGNED_OUTPUT           = 23,
	CMD_ADD                       = 1,
	CMD_SUBSRACT                  = 2,
	CMD_MULTIPLICATION            = 3,
	CMD_DIVISION                  = 4,
	CMD_UNSIGNED_MULTIPLICATION   = 19,
	CMD_UNSIGNED_DIVISION         = 20,
	CMD_ASSIGMENT                 = 0,
	CMD_GOTO                      = 128,
	CMD_LESS                      = 131,
	CMD_GREATER                   = 133,
	CMD_LESS_OR_EQUAL             = 134,
	CMD_GREATER_OR_EQUAL          = 132,
	CMD_EQUAL                     = 129,
	CMD_NOT_EQUAL                 = 130,
	CMD_UNSIGNED_LESS             = 147,
	CMD_UNSIGNED_GREATER          = 149,
	CMD_UNSIGNED_LESS_OR_EQUAL    = 150,
	CMD_UNSIGNED_GREATER_OR_EQUAL = 148,
	CMD_STOP                      = 153;


void Lm2::execute_the_program(){
	while (true){
		switch (this->memory.get(this->current_address)->get_value()){
		case CMD_COMPARE:
			this->perfom_flag_determination();
			break;
		case CMD_INPUT:
			this->perform_input_operation(LmCommands::input);
			break;
		case CMD_OUTPUT:
			this->perform_output_operation(LmCommands::output);
			break;
		case CMD_UNSIGNED_INPUT:
			this->perform_input_operation(LmCommands::unsigned_input);
			break;
		case CMD_UNSIGNED_OUTPUT:
			this->perform_output_operation(LmCommands::unsigned_output);
			break;
		case CMD_ADD:
			this->perform_arithmetic_operation(LmCommands::add,CMD_ADD);
			break;
		case CMD_SUBSRACT:
			this->perform_arithmetic_operation(LmCommands::subsract, CMD_SUBSRACT);
			break;
		case CMD_MULTIPLICATION:
			this->perform_arithmetic_operation(LmCommands::multiplication,CMD_MULTIPLICATION);
			break;
		case CMD_DIVISION:
			this->perform_arithmetic_operation(LmCommands::division,CMD_DIVISION);
			break;
		case CMD_UNSIGNED_MULTIPLICATION:
			this->perform_arithmetic_operation(LmCommands::unsigned_multiplication,CMD_UNSIGNED_MULTIPLICATION);
			break;
		case CMD_UNSIGNED_DIVISION:
			this->perform_arithmetic_operation(LmCommands::unsigned_division,CMD_UNSIGNED_DIVISION);
			break;
		case CMD_ASSIGMENT:
			perform_assignment_operation();
			break;
		case CMD_GOTO:
			this->perform_goto();
			break;
		case CMD_LESS:
			this->perform_comparison_operation(flags::LESS);
			break;
		case CMD_GREATER:
			this->perform_comparison_operation(flags::GREATER);
			break;
		case CMD_LESS_OR_EQUAL:
			this->perform_comparison_operation(flags::LESS_OR_EQUAL);
			break;
		case CMD_GREATER_OR_EQUAL:
			this->perform_comparison_operation(flags::GREATER_OR_EQUAL);
			break;
		case CMD_EQUAL:
			this->perform_comparison_operation(flags::EQUAL);
			break;
		case CMD_NOT_EQUAL:
			this->perform_comparison_operation(flags::N_EQUAL);
			break;
		case CMD_UNSIGNED_LESS:
			this->perform_comparison_operation(flags::INS_LESS);
			break;
		case CMD_UNSIGNED_GREATER:
			this->perform_comparison_operation(flags::INS_GREATER);
			break;
		case CMD_UNSIGNED_LESS_OR_EQUAL:
			this->perform_comparison_operation(flags::INS_LESS_OR_EQUAL);
			break;
		case CMD_UNSIGNED_GREATER_OR_EQUAL:
			this->perform_comparison_operation(flags::INS_GREATER_OR_EQUAL);
			break;
		case CMD_STOP:
			return;
			break;
		default:
			break;
		}
	}
}

void Lm2::clear_memory(){
	this->memory.clear();
}

void Lm2::init_variable(int position, std::string name){
	this->memory.set_name(position, name);
}

void Lm2::init_variable(int position, int value){
	this->memory.set(position, new Variable(value));
}

void Lm2::init_variable(int position, std::string name, int value){
	this->init_variable(position, name);
	this->init_variable(position, value);
}

void Lm2::set_program(const std::vector<std::vector<int> > &program){
	bool end = false;
	if (program.size() == 0)
		throw std::exception("The program is empty");

	if (program.front().front() >= 0 && program.front().front() <= MAX_MEMORY_SIZE)
		this->current_address=program.front().front();

	for (size_t i = 0; i < program.size(); i++){
		if (program[i].front() >= 0 && program[i].front() <= MAX_MEMORY_SIZE){
			try{
				this->memory.set(program[i].front(), new Lm2Command(program[i][1], program[i][2], program[i][3]));
				
					if (this->memory.get(program[i].front())->get_value() == CMD_STOP)
					 end = true;
				
			}
			catch (std::exception& exp){
				char* temp = new char[strlen(exp.what()) + 1];
				strcpy(temp, exp.what());
				throw InterpreterException(temp, i + 1);
			}

		}
		else{
			throw std::out_of_range("Memory bounds violation");
		}
	}
	
		if (!end)
		  throw std::exception("Lost end of program");

}

std::vector<std::vector<int> > Lm2::get_program()const{
	return std::vector<std::vector<int> >();
}

void Lm2::perform_arithmetic_operation(MemoryItem* (*func)(const MemoryItem*, const MemoryItem*),int type){
	MemoryItem* var_1 = new Variable(this->get_value_operand(this->current_address,1));
	MemoryItem* var_2 = new Variable(this->get_value_operand(this->current_address,2));
	this->memory.set(this->get_address_operand(this->current_address, 1), func(var_1, var_2));
	
	if (type == CMD_DIVISION || type == CMD_UNSIGNED_DIVISION)
		this->memory.set(this->get_address_operand(this->current_address, 1) + 1, LmCommands::module(var_1, var_2));
	
	this->current_address++;
}

void Lm2::perform_goto(){
	this->current_address = this->get_address_operand(this->current_address, 1);
}

void Lm2::perfom_flag_determination(){
	MemoryItem* var_1 = new Variable(this->get_value_operand(this->current_address, 1));
	MemoryItem* var_2 = new Variable(this->get_value_operand(this->current_address, 2));

	cmp_op[flags::EQUAL] = LmCommands::equal(var_1, var_2);
	cmp_op[flags::N_EQUAL] = LmCommands::not_equal(var_1, var_2);
	cmp_op[flags::LESS] = LmCommands::less(var_1, var_2);
	cmp_op[flags::GREATER] = LmCommands::greater(var_1, var_2);
	cmp_op[flags::LESS_OR_EQUAL] = LmCommands::less_or_equal(var_1, var_2);
	cmp_op[flags::GREATER_OR_EQUAL] = LmCommands::greater_or_equal(var_1, var_2);
	cmp_op[flags::INS_LESS] = LmCommands::unsigned_less(var_1, var_2);
	cmp_op[flags::INS_GREATER] = LmCommands::unsigned_greater(var_1, var_2);
	cmp_op[flags::INS_LESS_OR_EQUAL] = LmCommands::unsigned_less_or_equal(var_1, var_2);
	cmp_op[flags::INS_GREATER_OR_EQUAL] = LmCommands::unsigned_greater_or_equal(var_1, var_2);

	this->current_address++;
}
void Lm2::perform_comparison_operation(flags flag){
	this->current_address = cmp_op[flag] ? this->get_address_operand(this->current_address, 1) : this->current_address + 1;
}

void Lm2::perform_input_operation(void(*func)(MemoryItem*&, const std::string)){
	MemoryItem* var = new Variable();

	func(var, this->memory.get_name(this->get_address_operand(this->current_address, 1)));
	this->memory.set(this->get_address_operand(this->current_address, 1), var);

	this->current_address++;
}

void Lm2::perform_output_operation(void(*func)(const MemoryItem*, const std::string)){
	MemoryItem* var = new Variable(this->get_value_operand(this->current_address, 1));

	func(var, this->memory.get_name(this->get_address_operand(this->current_address, 1)));

	this->current_address++;
}

void Lm2::perform_assignment_operation(){
	int in_position = this->memory.get(this->current_address)->get().at(1);
	int value = this->get_value_operand(this->current_address, 2);

	MemoryItem* var = new Variable(value);
	
	this->memory.set(in_position, var);
	this->current_address++;
}

int Lm2::get_value_operand(int position_in_memory, int number_of_operands){
	if (position_in_memory < 0 || position_in_memory > MAX_MEMORY_SIZE)
		throw std::out_of_range("Memory bounds violation");

	if (number_of_operands < 0 || number_of_operands > 2)
		throw std::length_error("Invalid numbers of operands!");

	return this->memory.get(this->memory.get(position_in_memory)->get().at(number_of_operands))->get_value();
}

int Lm2::get_address_operand(int position_in_memory, int number_of_operands){
	if (position_in_memory < 0 || position_in_memory > MAX_MEMORY_SIZE)
		throw std::out_of_range("Memory bounds violation");

	if (number_of_operands < 0 || number_of_operands > 2)
		throw std::length_error("Invalid numbers of operands!");

	return this->memory.get(position_in_memory)->get().at(number_of_operands);
}

File: Windows - GUI CS/LM2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace LM_GUI_UP
{
    class LM2: LM, ILMOperations
    {
        private List<bool> comparisonOperations;
        public enum ComparisonFlag { EQUAL, N_EQUAL, LESS, GREATER, LESS_OR_EQUAL, GREATER_OR_EQUAL, INS_LESS, INS_GREATER, INS_LESS_OR_EQUAL, INS_GREATER_OR_EQUAL };
        #region Commands defines
            const int 
                CMD_COMPARE                   = 5,
	            CMD_INPUT                     = 6,
	            CMD_OUTPUT                    = 7,
	            CMD_UNSIGNED_INPUT            = 22,
	            CMD_UNSIGNED_OUTPUT           = 23,
	            CMD_ADD                       = 1,
	            CMD_SUBSTRACT                  = 2,
	            CMD_MULTIPLICATION            = 3,
	            CMD_DIVISION                  = 4,
	            CMD_UNSIGNED_MULTIPLICATION   = 19,
	            CMD_UNSIGNED_DIVISION         = 20,
	            CMD_ASSIGMENT                 = 0,
	            CMD_GOTO                      = 128,
	            CMD_LESS                      = 131,
	            CMD_GREATER                   = 133,
	            CMD_LESS_OR_EQUAL             = 134,
	            CMD_GREATER_OR_EQUAL          = 132,
	            CMD_EQUAL                     = 129,
	            CMD_NOT_EQUAL                 = 130,
	            CMD_UNSIGNED_LESS             = 147,
	            CMD_UNSIGNED_GREATER          = 149,
	            CMD_UNSIGNED_LESS_OR_EQUAL    = 150,
	            CMD_UNSIGNED_GREATER_OR_EQUAL = 148,
	            CMD_STOP                      = 153;
        #endregion

        public LM2():base(){
            comparisonOperations = new List<bool>(10);
        }

        #region Main methods
        
       /* public override void SetProgram(List<List<int > > parsedProgramText)
        {
            if (parsedProgramText.Count == 0)
                throw new Exception("");
            //The program is empty

            if(parsedProgramText[0][0] >= 0 && parsedProgramText[0][0] <= MAX_MEMORY_SIZE)
                currentAddress = parsedProgramText[0][0];

            for(int i = 0; i < parsedProgramText.Count; i++){
                if(parsedProgramText[i][0] >= 0 && parsedProgramText[i][0] <= MAX_MEMORY_SIZE){
                    InitialiseCommand(parsedProgramText[i]);
                }
                else
                {
                    throw new Exception();
                    //Runtime: Memory bounds violation
                }
            }
            if (!end)
                throw new Exception();
            //Runtime: Lost end of program

        }
        */

        protected override void InitialiseCommand(List<int> command)
        {
            try
            {
                CheckCommandFormat(command);

                this.ramMemory.SetCell(command[0], new LM2Command(command[1],
                    command[2], command[3]));
            }
            catch (Exception exp)
            {
                //char* temp = new char[strlen(exp.what()) + 1];
                //strcpy(temp, exp.what());
                //throw InterpreterException(temp, i + 1);
            }
        }

        private void CheckCommandFormat(List<int> command)
        {
            if (command.Count < 4)
                throw new Exception("Too few fields");
            else if (command.Count > 4)
                throw new Exception("Too many fields");
        }
        public override void DoOneStep(RichTextBox output)
        {
            if (this.endFlag)
                throw new Exception();
            //Program ended
            if (this.inputFlag)
                throw new Exception();
            //The value isn't entered
            switch (this.GetCurrentCommandNumber())
            {
                case CMD_COMPARE:
                    PerformFlagDetermination();
                    break;

                case CMD_OUTPUT:
                    PerformOutputOperation(LMCommands.Output, output);
                    break;
                case CMD_INPUT:
                    this.inputFlag = true;
                    break;
                case CMD_UNSIGNED_OUTPUT:
                    PerformOutputOperation(LMCommands.UnsignedOuput, output);
                    break;
                case CMD_UNSIGNED_INPUT:
                    this.inputFlag = true;
                    break;

                case CMD_ADD:
                    PerformArithmeticOperation(LMCommands.Add);
                    break;
                case CMD_SUBSTRACT:
                    PerformArithmeticOperation(LMCommands.Substract);
                    break;
                case CMD_MULTIPLICATION:
                    PerformArithmeticOperation(LMCommands.Multiplication);
                    break;
                case CMD_DIVISION:
                    PerformArithmeticOperation(LMCommands.Division);
                    break;
                case CMD_UNSIGNED_MULTIPLICATION:
                    PerformArithmeticOperation(LMCommands.UnsignedMultiplication);
                    break;
                case CMD_UNSIGNED_DIVISION:
                    PerformArithmeticOperation(LMCommands.UnsignedDivision);
                    break;

                case CMD_ASSIGMENT:
                    PerformAssignmentOperation();
                    break;

                case CMD_GOTO:
                    PerformGoToOperation();
                    break;

                case CMD_LESS:
                    PerformComparisonOperation(ComparisonFlag.LESS);
                    break;
                case CMD_GREATER:
                    PerformComparisonOperation(ComparisonFlag.GREATER);
                    break;
                case CMD_LESS_OR_EQUAL:
                    PerformComparisonOperation(ComparisonFlag.LESS_OR_EQUAL);
                    break;
                case CMD_GREATER_OR_EQUAL:
                    PerformComparisonOperation(ComparisonFlag.GREATER_OR_EQUAL);
                    break;
                case CMD_EQUAL:
                    PerformComparisonOperation(ComparisonFlag.EQUAL);
                    break;
                case CMD_NOT_EQUAL:
                    PerformComparisonOperation(ComparisonFlag.N_EQUAL);
                    break;

                case CMD_UNSIGNED_LESS:
                    PerformComparisonOperation(ComparisonFlag.INS_LESS);
                    break;
                case CMD_UNSIGNED_GREATER:
                    PerformComparisonOperation(ComparisonFlag.INS_GREATER);
                    break;
                case CMD_UNSIGNED_LESS_OR_EQUAL:
                    PerformComparisonOperation(ComparisonFlag.INS_LESS_OR_EQUAL);
                    break;
                case CMD_UNSIGNED_GREATER_OR_EQUAL:
                    PerformComparisonOperation(ComparisonFlag.INS_GREATER_OR_EQUAL);
                    break;
                case CMD_STOP:
                    this.endFlag = true;
                    break;
                default:
                    new Exception();
                    break;
            }
        }

        #endregion

        #region Axuiliary methods
        protected override void GoToNextAddress()
        {
            this.currentAddress++;
        }
        public override int GetValueOperand(int positionInMemory, int numberOfOperand)
        {
            if (positionInMemory < 0 || positionInMemory > MAX_MEMORY_SIZE)
                throw new Exception();
            //Memory bounds violation

            if (numberOfOperand < 0 || numberOfOperand > 2)
                throw new Exception();
            //Invalid number of operand
            return this.ramMemory.GetCell(this.ramMemory.GetCell(positionInMemory).GetItem()[numberOfOperand]).GetValue();
        }
        public override int GetAddressOperand(int positionInMemory, int numberOfOperand)
        {

            if (positionInMemory < 0 || positionInMemory > MAX_MEMORY_SIZE)
                throw new Exception();
            //Memory bounds violation

            if (numberOfOperand < 0 || numberOfOperand > 2)
                throw new Exception();
            //Invalid number of operand

            return this.ramMemory.GetCell(positionInMemory).GetItem()[numberOfOperand];
        }
        public override void InitVariable(int position, string name)
        {
            this.ramMemory.SetName(position, name);
        }
        public override void InitVariable(int position, int value)
        {
            this.ramMemory.SetCell(position, new Variable(value));
        }
        public override void InitVariable(int position, string name, int value)
        {
            InitVariable(position, name);
            InitVariable(position, value);
        }
        #endregion

        #region ILMOperations realisation
        public void PerformArithmeticOperation(Func<MemoryItem, MemoryItem, MemoryItem> arithmeticDelegate)
        {
            MemoryItem var1 = new Variable(GetValueOperand(this.currentAddress, 1));
            MemoryItem var2 = new Variable(GetValueOperand(this.currentAddress, 2));

            this.ramMemory.SetCell(GetAddressOperand(this.currentAddress, 1), arithmeticDelegate(var1, var2));
            //possible mistake with module
            if (arithmeticDelegate.Method.Name == "Division" || arithmeticDelegate.Method.Name == "UnsignedDivision")
                this.ramMemory.SetCell(GetAddressOperand(this.currentAddress, 1) + 1, LMCommands.Module(var1, var2));

            GoToNextAddress();
        }
        public void PerformFlagDetermination()
        {
            MemoryItem var1 = new Variable(GetValueOperand(this.currentAddress, 1));
            MemoryItem var2 = new Variable(GetValueOperand(this.currentAddress, 2));


            comparisonOperations[(int)ComparisonFlag.EQUAL] = LMCommands.Equal(var1, var2);
            comparisonOperations[(int)ComparisonFlag.N_EQUAL] = LMCommands.NotEqual(var1, var2);
            comparisonOperations[(int)ComparisonFlag.LESS] = LMCommands.Less(var1, var2);
            comparisonOperations[(int)ComparisonFlag.GREATER] = LMCommands.Greater(var1, var2);
            comparisonOperations[(int)ComparisonFlag.LESS_OR_EQUAL] = LMCommands.LessOrEqual(var1, var2);
            comparisonOperations[(int)ComparisonFlag.GREATER_OR_EQUAL] = LMCommands.GreaterOrEqual(var1, var2);
            comparisonOperations[(int)ComparisonFlag.INS_LESS] = LMCommands.UnsignedLess(var1, var2);
            comparisonOperations[(int)ComparisonFlag.INS_GREATER] = LMCommands.UnsignedGreater(var1, var2);
            comparisonOperations[(int)ComparisonFlag.INS_LESS_OR_EQUAL] = LMCommands.UnsignedLessOrEqual(var1, var2);
            comparisonOperations[(int)ComparisonFlag.INS_GREATER_OR_EQUAL] = LMCommands.UnsignedGreaterOrEqual(var1, var2);

            GoToNextAddress();
        }
        public void PerformComparisonOperation(ComparisonFlag flag)
        {
            this.currentAddress = comparisonOperations[(int)flag] ? GetAddressOperand(this.currentAddress, 1) : this.currentAddress + 1;
        }
        public void PerformInputOperation(int value)
        {
            MemoryItem var = new Variable(value);

            this.ramMemory.SetCell(GetAddressOperand(this.currentAddress, 1), var);

            this.inputFlag = false;
            GoToNextAddress();

        }
        public void PerformOutputOperation(Action<RichTextBox, MemoryItem, string> outputDelegate,
            RichTextBox output)
        {
            MemoryItem var = new Variable(this.GetValueOperand(this.currentAddress, 1));

            outputDelegate(output, var, this.ramMemory.GetName(GetAddressOperand(this.currentAddress, 1)));

            GoToNextAddress();
        }
        public void PerformAssignmentOperation()
        {
            int inPosition = GetAddressOperand(this.currentAddress, 1);
            int value = GetValueOperand(this.currentAddress, 2);

            MemoryItem var = new Variable(value);

            this.ramMemory.SetCell(inPosition, var);

            GoToNextAddress();
        }
        public void PerformGoToOperation()
        {
            this.currentAddress = this.GetAddressOperand(this.currentAddress, 1);
        }

        #endregion
    }
}
